frontend/ui/screensaver.lua

```diff
@@ -60,93 +38,227 @@ local function getRandomImage(dir)
     local doc_settings = DocSettings:open(lastfile)
-    if image then
-        local img_widget = ImageWidget:new{
-            image = image,
-            height = Screen:getHeight(),
-            width = Screen:getWidth(),
-            scale_factor = doc_settings:readSetting("proportional_screensaver") and 0 or nil,
-        }
-        return createWidgetFromImage(img_widget)
+    local exclude_ss = doc_settings:readSetting("exclude_screensaver")
+    doc_settings:close()
+    return exclude_ss or false
+end
+
+function Screensaver:setMessage()
+    local InputDialog = require("ui/widget/inputdialog")
+    local screensaver_message = G_reader_settings:readSetting("screensaver_message")
+    if screensaver_message == nil then
+        screensaver_message = default_screensaver_message
     end
+    self.input_dialog = InputDialog:new{
+        title = "Screensaver message",
+        input = screensaver_message,
+        buttons = {
+            {
+                {
+                    text = _("Cancel"),
+                    callback = function()
+                        UIManager:close(self.input_dialog)
+                    end,
+                },
+                {
+                    text = _("Set message"),
+                    is_enter_default = true,
+                    callback = function()
+                        G_reader_settings:saveSetting("screensaver_message", self.input_dialog:getInputText())
+                        UIManager:close(self.input_dialog)
+                    end,
+                },
+            },
+        },
+    }
+    self.input_dialog:onShowKeyboard()
+    UIManager:show(self.input_dialog)
 end
 
-function Screensaver:show(kind, default_msg)
-    logger.dbg("show screensaver")
-    local InfoMessage = require("ui/widget/infomessage")
-    local screensaver_settings = G_reader_settings:readSetting(kind .. "_screensaver") or {}
-    -- first check book cover image, on by default
-    local screensaver_last_book = screensaver_settings.use_last_file or
-        G_reader_settings:readSetting("use_lastfile_as_screensaver")
-    if screensaver_last_book == nil or screensaver_last_book then
+function Screensaver:show()
+    if self.left_msg then
+        UIManager:close(self.left_msg)
+        self.left_msg = nil
+    end
+    local screensaver_type = G_reader_settings:readSetting("screensaver_type")
+    local widget = nil
+    local background = Blitbuffer.COLOR_WHITE
+    if screensaver_type == "disable" or screensaver_type == nil then
+        return
+    end
+    if screensaver_type == "cover" then
         local lastfile = G_reader_settings:readSetting("lastfile")
-        if lastfile then
-            local doc_settings = DocSettings:open(lastfile)
-            local exclude = doc_settings:readSetting("exclude_screensaver")
-            if not exclude then
-                self.left_msg = self:getCoverImage(lastfile)
-            end
+        local doc_settings = DocSettings:open(lastfile)
```

Comment:
If this lastfile book had its .sdr purged (user reads, goes back to filemanager, Purge .sdr, goes to screensaver), this will recreate the .sdr. `DocSettings:open(lastfile)` creates it. We could use something like:
```lua
local exclude = false -- consider it not excluded if there's no docsetting
if DocSettings:hasSidecarFile(lastfile) then
  local doc_settings = DocSettings:open(lastfile)
  exclude = doc_settings:readSetting("exclude_screensaver")
end
```

Also probably needs to be done too in your Screensaver:excluded() above where the same thing is done (may be you could use this function here to avoid duplicate code, haven't checked more if they do the same thing).